org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/rewrite/ImportRewrite.java

```diff
@@ -337,15 +341,114 @@ public static ImportRewrite create(ICompilationUnit cu, boolean restoreExistingI
 			throw new IllegalArgumentException("Compilation unit must not be null"); //$NON-NLS-1$
 		}
 		List existingImport= null;
+		Map<String, List<String>> moduleEntries= null;
 		if (restoreExistingImports) {
 			existingImport= new ArrayList();
+			moduleEntries= new HashMap<>();
 			IImportDeclaration[] imports= cu.getImports();
+			CompilationUnit compilationUnit= null;
 			for (IImportDeclaration curr : imports) {
-				char prefix= Flags.isStatic(curr.getFlags()) ? STATIC_PREFIX : NORMAL_PREFIX;
-				existingImport.add(prefix + curr.getElementName());
+				char prefix= Flags.isStatic(curr.getFlags()) ? STATIC_PREFIX :
+					Flags.isModule(curr.getFlags()) ? MODULE_PREFIX : NORMAL_PREFIX;
+				String currName= curr.getElementName();
+				if (currName.endsWith(".*") && prefix == MODULE_PREFIX) { //$NON-NLS-1$
+					currName= currName.substring(0, currName.length() - 2);
+				}
+				existingImport.add(prefix + currName);
+				if (Flags.isModule(curr.getFlags())) {
+					List<String> packageNames= new ArrayList<>();
+					if (compilationUnit == null) {
+						compilationUnit= convertICUtoCU(cu);
+					}
+					if (compilationUnit != null) {
+						List<ImportDeclaration> astImports= compilationUnit.imports();
+						ImportDeclaration foundModuleImport= null;
+						for (ImportDeclaration astImport : astImports) {
+							if (Modifier.isModule(astImport.getModifiers()) && astImport.getName().getFullyQualifiedName().equals(currName)) {
+								foundModuleImport= astImport;
+								break;
+							}
+						}
+						if (foundModuleImport != null) {
+							IBinding moduleImportBinding= foundModuleImport.resolveBinding();
+							if (moduleImportBinding instanceof IModuleBinding moduleBinding) {
+								packageNames= getPackageNamesForModule(moduleBinding, compilationUnit);
+							}
+						}
+					}
+					String name= curr.getElementName();
+					if (name.endsWith(".*")) { //$NON-NLS-1$
+						name= name.substring(0, name.length() - 2);
+					}
+					moduleEntries.put(name, packageNames);
+				}
+			}
+		}
+		return new ImportRewrite(cu, null, existingImport, moduleEntries);
+	}
+
+	/**
+	 * Calculate the list of package names that are exported by a module including
+	 * direct and transitive imports.
+	 *
+	 * @param binding - module binding
+	 * @param cu - compilation unit
+	 * @return a list of package names that are exported by the given binding
+	 *
+	 * @since 3.43
+	 */
+	public static List<String> getPackageNamesForModule(IModuleBinding binding, CompilationUnit cu) {
+		Set<IModuleBinding> modules= new HashSet<>();
+		String currentModuleName= null;
+		try {
+			IModuleDescription modDesc= cu.getJavaElement().getJavaProject().getModuleDescription();
+			currentModuleName= modDesc.getElementName();
```

Comment:
`null` is a legal return value from `getModuleDescription()`.

If that happens, we're in the unnamed module, and hence we cannot leverage any qualified exports (`to`).

Currently, 3 new tests in JDT/UI throw NullPointerException at this location.